File: Odysseus and the Bag of Winds/claude2.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    const ll INF = LLONG_MAX / 2;
    
    int T;
    cin >> T;
    
    while (T--) {
        int M, A;
        cin >> M >> A;
        int Aidx = A - 1;
        
        vector<vector<int>> C(M, vector<int>(M));
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < M; j++) {
                cin >> C[i][j];
            }
        }
        
        int fullMask = 1 << M;
        vector<vector<vector<ll>>> dp(fullMask, vector<vector<ll>>(M, vector<ll>(2, INF)));
        
        dp[1][0][0] = 0;
        
        for (int mask = 1; mask < fullMask; mask++) {
            if (!(mask & 1)) continue;
            
            bool aVisited = (mask & (1 << Aidx)) != 0;
            
            for (int last = 0; last < M; last++) {
                if (!(mask & (1 << last))) continue;
                
                for (int bagState = 0; bagState < 2; bagState++) {
                    ll currentCost = dp[mask][last][bagState];
                    if (currentCost >= INF) continue;
                    
                    for (int next = 0; next < M; next++) {
                        if (mask & (1 << next)) continue;
                        
                        int newMask = mask | (1 << next);
                        
                        ll normalCost = currentCost + C[last][next];
                        if (normalCost < dp[newMask][next][bagState]) {
                            dp[newMask][next][bagState] = normalCost;
                        }
                        
                        if (bagState == 0 && aVisited) {
                            ll halvedCost = currentCost + C[last][next] / 2;
                            if (halvedCost < dp[newMask][next][1]) {
                                dp[newMask][next][1] = halvedCost;
                            }
                        }
                    }
                }
            }
        }
        
        int finalMask = fullMask - 1;
        ll answer = min(dp[finalMask][M-1][0], dp[finalMask][M-1][1]);
        cout << answer << "\n";
    }
    
    return 0;
}

File: Odysseus and the Bag of Winds/claude2.py
import sys

def solve():
    input_data = sys.stdin.read().split()
    pos = 0
    T = int(input_data[pos]); pos += 1
    
    INF = float('inf')
    results = []
    
    for _ in range(T):
        M = int(input_data[pos]); pos += 1
        A = int(input_data[pos]); pos += 1
        A_idx = A - 1
        
        C = [[0] * M for _ in range(M)]
        for i in range(M):
            for j in range(M):
                C[i][j] = int(input_data[pos])
                pos += 1
        
        FULL_MASK = 1 << M
        
        dp = [[[INF, INF] for _ in range(M)] for _ in range(FULL_MASK)]
        dp[1][0][0] = 0
        
        for mask in range(1, FULL_MASK):
            if not (mask & 1):
                continue
            
            a_visited = bool(mask & (1 << A_idx))
            
            for last_node in range(M):
                if not (mask & (1 << last_node)):
                    continue
                
                for bag_state in range(2):
                    current_cost = dp[mask][last_node][bag_state]
                    if current_cost == INF:
                        continue
                    
                    for next_node in range(M):
                        if mask & (1 << next_node):
                            continue
                        
                        new_mask = mask | (1 << next_node)
                        
                        normal_cost = current_cost + C[last_node][next_node]
                        if normal_cost < dp[new_mask][next_node][bag_state]:
                            dp[new_mask][next_node][bag_state] = normal_cost
                        
                        if bag_state == 0 and a_visited:
                            halved_cost = current_cost + C[last_node][next_node] // 2
                            if halved_cost < dp[new_mask][next_node][1]:
                                dp[new_mask][next_node][1] = halved_cost
        
        final_mask = FULL_MASK - 1
        answer = min(dp[final_mask][M-1][0], dp[final_mask][M-1][1])
        results.append(str(answer))
    
    print('\n'.join(results))

if __name__ == "__main__":
    solve()